syntax = "proto2";

message RequestVoteArgs {
    required uint32 term = 1;
    required int32 candidate_id = 2;
    required uint32 last_log_index = 3;
    required uint32 last_log_term = 4;
}

message RequestVoteReply {
    required uint32 term = 1;
    required bool vote_granted = 2;
}

message LogEntry {
    required uint32 term = 1;
    required uint32 index = 2;
    required bytes command = 3;
}

message AppendEntriesArgs {
    required uint32 term = 1;
    required int32 leader_id = 2;
    required uint32 prev_log_index = 3;
    required uint32 prev_log_term = 4;
    repeated LogEntry entries = 5;
    required uint32 leader_commit = 6;
}

message AppendEntriesReply {
    required uint32 term = 1;
    required bool success = 2;
    required uint32 conflict_index = 3;
    required uint32 conflict_term = 4;
}
